File: services/agents/app/routers/llm_connections.py
"""Strict project-scoped LLM provider connection and model discovery API."""

from __future__ import annotations

from datetime import datetime
from typing import Literal
from uuid import UUID

from fastapi import APIRouter, HTTPException, Path, Query, Request, status
from pydantic import BaseModel, ConfigDict, Field, SecretStr, field_validator

from app.auth import Principal, require_project
from app.llm.provider_catalog import (
    ProviderDiscoveryError,
    ProviderModel,
    discover_models,
)
from app.store.llm_connections import (
    ConnectionMetadata,
    LlmConnectionAssignmentConflictError,
    LlmConnectionAuthorizationError,
    LlmConnectionConflictError,
    LlmConnectionNotFoundError,
    ProjectConnectionStore,
)
from app.store.llm_credentials import CredentialStoreError


ProviderPath = Literal["openai", "anthropic", "google", "xai"]
PROJECT_PATTERN = r"^[A-Za-z0-9]{1,64}$"

router = APIRouter(prefix="/v1/agents/llm-connections", tags=["agents"])


class PutConnectionRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    project_id: str = Field(pattern=PROJECT_PATTERN)
    api_key: SecretStr = Field(min_length=1, max_length=16_384)
    version: int = Field(ge=0)

    @field_validator("api_key")
    @classmethod
    def validate_api_key_size(cls, value: SecretStr) -> SecretStr:
        if len(value.get_secret_value().encode("utf-8")) > 16_384:
            raise ValueError("api_key must not exceed 16384 UTF-8 bytes")
        return value


class RefreshConnectionRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    project_id: str = Field(pattern=PROJECT_PATTERN)
    version: int = Field(ge=1)


class RevokeConnectionRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    project_id: str = Field(pattern=PROJECT_PATTERN)
    version: int = Field(ge=1)
    reason: str = Field(min_length=1, max_length=2_000)

    @field_validator("reason")
    @classmethod
    def validate_reason(cls, value: str) -> str:
        if value != value.strip() or "\r" in value or "\n" in value:
            raise ValueError(
                "reason must not contain surrounding whitespace or line breaks"
            )
        return value


class ProviderModelResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    schema_version: Literal["llm_provider_model@1"]
    provider: ProviderPath
    model_id: str = Field(pattern=r"^[A-Za-z0-9][A-Za-z0-9._:/-]{0,127}$")
    display_name: str
    supported_tiers: list[Literal["fast", "reasoning"]]
    catalog_version: str
    data_residency: Literal["ca", "us", "eu", "global"]
    allowed_data_classifications: list[
        Literal["public", "internal", "confidential", "restricted"]
    ]
    pricing_status: Literal["operator_review_required"]


class ConnectionSummaryResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    schema_version: Literal["llm_provider_connection@1"] = (
        "llm_provider_connection@1"
    )
    project_id: str = Field(pattern=PROJECT_PATTERN)
    provider: ProviderPath
    version: int = Field(ge=1)
    state: Literal["active", "revoked"]
    catalog_version: str
    validated_at: datetime
    created_at: datetime
    updated_at: datetime
    revoked_at: datetime | None
    model_count: int = Field(ge=0, le=1_000)


class ConnectionDetailResponse(ConnectionSummaryResponse):
    models: list[ProviderModelResponse]


class ConnectionListResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    schema_version: Literal["llm_provider_connection_list@1"] = (
        "llm_provider_connection_list@1"
    )
    project_id: str = Field(pattern=PROJECT_PATTERN)
    connections: list[ConnectionSummaryResponse]


class ModelInventoryResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    schema_version: Literal["llm_provider_model_inventory@1"] = (
        "llm_provider_model_inventory@1"
    )
    project_id: str = Field(pattern=PROJECT_PATTERN)
    provider: ProviderPath
    connection_version: int = Field(ge=1)
    models: list[ProviderModelResponse]


def _summary(connection: ConnectionMetadata) -> ConnectionSummaryResponse:
    return ConnectionSummaryResponse(
        project_id=connection.project_id,
        provider=connection.provider,
        version=connection.version,
        state=connection.state,
        catalog_version=connection.catalog_version,
        validated_at=connection.validated_at,
        created_at=connection.created_at,
        updated_at=connection.updated_at,
        revoked_at=connection.revoked_at,
        model_count=connection.model_count,
    )


def _model(model: ProviderModel) -> ProviderModelResponse:
    return ProviderModelResponse(
        schema_version=model.schema_version,
        provider=model.provider,
        model_id=model.model_id,
        display_name=model.display_name,
        supported_tiers=list(model.supported_tiers),
        catalog_version=model.catalog_version,
        data_residency=model.data_residency,
        allowed_data_classifications=list(model.allowed_data_classifications),
        pricing_status=model.pricing_status,
    )


def _store(request: Request) -> ProjectConnectionStore:
    return request.app.state.llm_connection_store


def _mutation_actor(request: Request, project_id: str) -> UUID:
    principal: Principal = request.state.principal
    if principal.project_id != project_id:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Credential is not authorized for this project",
        )
    if principal.actor_user_id is None:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="A current human session is required",
        )
    return UUID(principal.actor_user_id)


def _provider_error(exc: ProviderDiscoveryError) -> HTTPException:
    return HTTPException(
        status_code=exc.status_code,
        detail={"code": exc.code, "message": str(exc)},
    )


def _store_error(exc: Exception) -> HTTPException:
    if isinstance(exc, LlmConnectionAuthorizationError):
        return HTTPException(status_code=403, detail=str(exc))
    if isinstance(exc, LlmConnectionNotFoundError):
        return HTTPException(status_code=404, detail=str(exc))
    if isinstance(
        exc,
        (
            LlmConnectionConflictError,
            LlmConnectionAssignmentConflictError,
        ),
    ):
        return HTTPException(status_code=409, detail=str(exc))
    return HTTPException(
        status_code=503,
        detail="LLM connection storage is unavailable",
    )


async def _preflight_mutation(
    request: Request,
    project_id: str,
) -> tuple[ProjectConnectionStore, UUID]:
    store = _store(request)
    actor_user_id = _mutation_actor(request, project_id)
    try:
        await store.assert_mutation_authority(project_id, actor_user_id)
    except LlmConnectionAuthorizationError as exc:
        raise _store_error(exc) from exc
    return store, actor_user_id


@router.get("", response_model=ConnectionListResponse)
async def list_connections(
    request: Request,
    project_id: str = Query(..., pattern=PROJECT_PATTERN),
) -> ConnectionListResponse:
    require_project(request, project_id, "agents:read")
    try:
        connections = await _store(request).list(project_id)
    except Exception as exc:
        raise _store_error(exc) from exc
    return ConnectionListResponse(
        project_id=project_id,
        connections=[_summary(connection) for connection in connections],
    )


@router.put("/{provider}", response_model=ConnectionDetailResponse)
async def put_connection(
    body: PutConnectionRequest,
    request: Request,
    provider: ProviderPath = Path(...),
) -> ConnectionDetailResponse:
    store, actor_user_id = await _preflight_mutation(request, body.project_id)
    api_key = body.api_key.get_secret_value()
    try:
        models = await discover_models(provider, api_key)
        connection = await store.put(
            body.project_id,
            provider,
            api_key,
            models,
            expected_version=body.version,
            actor_user_id=actor_user_id,
        )
    except ProviderDiscoveryError as exc:
        raise _provider_error(exc) from exc
    except (
        CredentialStoreError,
        LlmConnectionAuthorizationError,
        LlmConnectionConflictError,
        LlmConnectionAssignmentConflictError,
        LlmConnectionNotFoundError,
    ) as exc:
        raise _store_error(exc) from exc
    except Exception as exc:
        raise _store_error(exc) from exc
    finally:
        api_key = ""
    return ConnectionDetailResponse(
        **_summary(connection).model_dump(),
        models=[_model(model) for model in models],
    )


@router.get("/{provider}/models", response_model=ModelInventoryResponse)
async def get_models(
    request: Request,
    provider: ProviderPath = Path(...),
    project_id: str = Query(..., pattern=PROJECT_PATTERN),
) -> ModelInventoryResponse:
    require_project(request, project_id, "agents:read")
    try:
        connection, models = await _store(request).get_active_with_models(
            project_id,
            provider,
        )
    except (
        LlmConnectionNotFoundError,
        LlmConnectionConflictError,
    ) as exc:
        raise _store_error(exc) from exc
    except Exception as exc:
        raise _store_error(exc) from exc
    return ModelInventoryResponse(
        project_id=project_id,
        provider=provider,
        connection_version=connection.version,
        models=[_model(model) for model in models],
    )


@router.post(
    "/{provider}/refresh-models",
    response_model=ConnectionDetailResponse,
)
async def refresh_models(
    body: RefreshConnectionRequest,
    request: Request,
    provider: ProviderPath = Path(...),
) -> ConnectionDetailResponse:
    store, actor_user_id = await _preflight_mutation(request, body.project_id)
    api_key = ""
    credential = None
    try:
        credential = await store.credential_for_refresh(
            body.project_id,
            provider,
            expected_version=body.version,
        )
        api_key = credential.api_key
        models = await discover_models(provider, api_key)
        connection = await store.refresh(
            body.project_id,
            provider,
            models,
            expected_version=body.version,
            expected_credential_id=credential.credential_id,
            actor_user_id=actor_user_id,
        )
    except ProviderDiscoveryError as exc:
        raise _provider_error(exc) from exc
    except (
        CredentialStoreError,
        LlmConnectionAuthorizationError,
        LlmConnectionConflictError,
        LlmConnectionAssignmentConflictError,
        LlmConnectionNotFoundError,
    ) as exc:
        raise _store_error(exc) from exc
    except Exception as exc:
        raise _store_error(exc) from exc
    finally:
        api_key = ""
        credential = None
    return ConnectionDetailResponse(
        **_summary(connection).model_dump(),
        models=[_model(model) for model in models],
    )


@router.post("/{provider}/revoke", response_model=ConnectionSummaryResponse)
async def revoke_connection(
    body: RevokeConnectionRequest,
    request: Request,
    provider: ProviderPath = Path(...),
) -> ConnectionSummaryResponse:
    store, actor_user_id = await _preflight_mutation(request, body.project_id)
    try:
        connection = await store.revoke(
            body.project_id,
            provider,
            expected_version=body.version,
            actor_user_id=actor_user_id,
            reason=body.reason,
        )
    except (
        CredentialStoreError,
        LlmConnectionAuthorizationError,
        LlmConnectionConflictError,
        LlmConnectionAssignmentConflictError,
        LlmConnectionNotFoundError,
    ) as exc:
        raise _store_error(exc) from exc
    except Exception as exc:
        raise _store_error(exc) from exc
    return _summary(connection)

File: services/admin/src/api/llmConnections.ts
import { z } from 'zod'

import { request, type ServiceConnection } from '@/api/http'

export const llmProviderSchema = z.enum(['openai', 'anthropic', 'google', 'xai'])
export const llmModelTierSchema = z.enum(['fast', 'reasoning'])

const projectIdSchema = z.string().regex(/^[A-Za-z0-9]{1,64}$/)
const timestampSchema = z.string().datetime({ offset: true })

export const llmProviderModelSchema = z
  .object({
    schema_version: z.literal('llm_provider_model@1'),
    provider: llmProviderSchema,
    model_id: z.string().regex(/^[A-Za-z0-9][A-Za-z0-9._:/-]{0,127}$/),
    display_name: z.string(),
    supported_tiers: z.array(llmModelTierSchema),
    catalog_version: z.string(),
    data_residency: z.enum(['ca', 'us', 'eu', 'global']),
    allowed_data_classifications: z.array(
      z.enum(['public', 'internal', 'confidential', 'restricted']),
    ),
    pricing_status: z.literal('operator_review_required'),
  })
  .strict()

const llmConnectionSummaryObjectSchema = z
  .object({
    schema_version: z.literal('llm_provider_connection@1'),
    project_id: projectIdSchema,
    provider: llmProviderSchema,
    version: z.number().int().min(1),
    state: z.enum(['active', 'revoked']),
    catalog_version: z.string(),
    validated_at: timestampSchema,
    created_at: timestampSchema,
    updated_at: timestampSchema,
    revoked_at: timestampSchema.nullable(),
    model_count: z.number().int().min(0).max(1_000),
  })
  .strict()

function validateConnectionState(
  connection: z.infer<typeof llmConnectionSummaryObjectSchema>,
  context: z.RefinementCtx,
): void {
  if (connection.state === 'active' && connection.revoked_at !== null) {
    context.addIssue({
      code: z.ZodIssueCode.custom,
      path: ['revoked_at'],
      message: 'active connections cannot have revoked_at',
    })
  }
  if (connection.state === 'revoked' && connection.revoked_at === null) {
    context.addIssue({
      code: z.ZodIssueCode.custom,
      path: ['revoked_at'],
      message: 'revoked connections require revoked_at',
    })
  }
}

export const llmConnectionSummarySchema =
  llmConnectionSummaryObjectSchema.superRefine(validateConnectionState)

export const llmConnectionDetailSchema = llmConnectionSummaryObjectSchema
  .extend({
    models: z.array(llmProviderModelSchema),
  })
  .strict()
  .superRefine((connection, context) => {
    validateConnectionState(connection, context)
    if (connection.model_count !== connection.models.length) {
      context.addIssue({
        code: z.ZodIssueCode.custom,
        path: ['model_count'],
        message: 'model_count must match models',
      })
    }
    connection.models.forEach((model, index) => {
      if (model.provider !== connection.provider) {
        context.addIssue({
          code: z.ZodIssueCode.custom,
          path: ['models', index, 'provider'],
          message: 'model provider must match connection provider',
        })
      }
    })
  })

export const llmConnectionListSchema = z
  .object({
    schema_version: z.literal('llm_provider_connection_list@1'),
    project_id: projectIdSchema,
    connections: z.array(llmConnectionSummarySchema),
  })
  .strict()
  .superRefine((list, context) => {
    const providers = new Set<LlmProvider>()
    list.connections.forEach((connection, index) => {
      if (connection.project_id !== list.project_id) {
        context.addIssue({
          code: z.ZodIssueCode.custom,
          path: ['connections', index, 'project_id'],
          message: 'connection project must match list project',
        })
      }
      if (providers.has(connection.provider)) {
        context.addIssue({
          code: z.ZodIssueCode.custom,
          path: ['connections', index, 'provider'],
          message: 'connection providers must be unique',
        })
      }
      providers.add(connection.provider)
    })
  })

export const llmModelInventorySchema = z
  .object({
    schema_version: z.literal('llm_provider_model_inventory@1'),
    project_id: projectIdSchema,
    provider: llmProviderSchema,
    connection_version: z.number().int().min(1),
    models: z.array(llmProviderModelSchema),
  })
  .strict()
  .superRefine((inventory, context) => {
    inventory.models.forEach((model, index) => {
      if (model.provider !== inventory.provider) {
        context.addIssue({
          code: z.ZodIssueCode.custom,
          path: ['models', index, 'provider'],
          message: 'model provider must match inventory provider',
        })
      }
    })
  })

export const putLlmConnectionRequestSchema = z
  .object({
    project_id: projectIdSchema,
    api_key: z.string().min(1).max(16_384),
    version: z.number().int().min(0),
  })
  .strict()
  .superRefine((body, context) => {
    if (new TextEncoder().encode(body.api_key).length > 16_384) {
      context.addIssue({
        code: z.ZodIssueCode.custom,
        path: ['api_key'],
        message: 'api_key must not exceed 16384 UTF-8 bytes',
      })
    }
  })

export const refreshLlmConnectionRequestSchema = z
  .object({
    project_id: projectIdSchema,
    version: z.number().int().min(1),
  })
  .strict()

export const revokeLlmConnectionRequestSchema = z
  .object({
    project_id: projectIdSchema,
    version: z.number().int().min(1),
    reason: z
      .string()
      .min(1)
      .max(2_000)
      .refine(
        (reason) =>
          reason === reason.trim() && !reason.includes('\r') && !reason.includes('\n'),
        'reason must not contain surrounding whitespace or line breaks',
      ),
  })
  .strict()

export type LlmProvider = z.infer<typeof llmProviderSchema>
export type LlmProviderModel = z.infer<typeof llmProviderModelSchema>
export type LlmConnectionSummary = z.infer<typeof llmConnectionSummarySchema>
export type LlmConnectionDetail = z.infer<typeof llmConnectionDetailSchema>
export type LlmConnectionList = z.infer<typeof llmConnectionListSchema>
export type LlmModelInventory = z.infer<typeof llmModelInventorySchema>

function isProviderCredentialRejection(body: unknown): boolean {
  if (typeof body !== 'object' || body === null || !('detail' in body)) return false
  const detail = body.detail
  return (
    typeof detail === 'object' &&
    detail !== null &&
    'code' in detail &&
    detail.code === 'invalid_key' &&
    'message' in detail &&
    typeof detail.message === 'string'
  )
}

function connectionPath(provider?: LlmProvider): string {
  const base = '/v1/agents/llm-connections'
  return provider ? `${base}/${provider}` : base
}

export function listLlmConnections(
  connection: ServiceConnection,
  projectId: string,
  signal?: AbortSignal,
): Promise<LlmConnectionList> {
  return request(connection, connectionPath(), {
    query: { project_id: projectId },
    signal,
    schema: llmConnectionListSchema,
  }).then((result) => {
    if (result.project_id !== projectId) {
      throw new Error('LLM connection list crossed project authority')
    }
    return result
  })
}

export function putLlmConnection(
  connection: ServiceConnection,
  provider: LlmProvider,
  projectId: string,
  apiKey: string,
  version: number,
): Promise<LlmConnectionDetail> {
  const body = putLlmConnectionRequestSchema.parse({
    project_id: projectId,
    api_key: apiKey,
    version,
  })
  return request(connection, connectionPath(provider), {
    method: 'PUT',
    body,
    schema: llmConnectionDetailSchema,
    // Suppress sign-out only for the canonical provider-auth rejection. A BFF
    // session 401 must still terminate the Admin session.
    redirectOnUnauthorized: (body) => !isProviderCredentialRejection(body),
  }).then((result) => {
    if (result.project_id !== projectId || result.provider !== provider) {
      throw new Error('LLM connection response crossed project authority')
    }
    return result
  })
}

export function getLlmModels(
  connection: ServiceConnection,
  provider: LlmProvider,
  projectId: string,
  signal?: AbortSignal,
): Promise<LlmModelInventory> {
  return request(connection, `${connectionPath(provider)}/models`, {
    query: { project_id: projectId },
    signal,
    schema: llmModelInventorySchema,
  }).then((result) => {
    if (result.project_id !== projectId || result.provider !== provider) {
      throw new Error('LLM model inventory crossed project authority')
    }
    return result
  })
}

export function refreshLlmModels(
  connection: ServiceConnection,
  provider: LlmProvider,
  projectId: string,
  version: number,
): Promise<LlmConnectionDetail> {
  const body = refreshLlmConnectionRequestSchema.parse({
    project_id: projectId,
    version,
  })
  return request(connection, `${connectionPath(provider)}/refresh-models`, {
    method: 'POST',
    body,
    schema: llmConnectionDetailSchema,
    // Refresh can expose the same provider-auth rejection as connect.
    redirectOnUnauthorized: (body) => !isProviderCredentialRejection(body),
  }).then((result) => {
    if (result.project_id !== projectId || result.provider !== provider) {
      throw new Error('LLM model refresh crossed project authority')
    }
    return result
  })
}

export function revokeLlmConnection(
  connection: ServiceConnection,
  provider: LlmProvider,
  projectId: string,
  version: number,
  reason: string,
): Promise<LlmConnectionSummary> {
  const body = revokeLlmConnectionRequestSchema.parse({
    project_id: projectId,
    version,
    reason,
  })
  return request(connection, `${connectionPath(provider)}/revoke`, {
    method: 'POST',
    body,
    schema: llmConnectionSummarySchema,
  }).then((result) => {
    if (result.project_id !== projectId || result.provider !== provider) {
      throw new Error('LLM connection revocation crossed project authority')
    }
    return result
  })
}
